tensorcircuit/abstractcircuit.py

```diff
@@ -0,0 +1,609 @@
+"""
+Methods for abstract circuits independent of nodes, edges and contractions
+"""
+# pylint: disable=invalid-name
+
+from typing import Any, Callable, Dict, List, Optional, Sequence, Union
+from functools import reduce
+from operator import add
+
+import numpy as np
+import tensornetwork as tn
+
+from . import gates
+from .cons import backend, dtypestr
+from .vis import qir2tex
+from .quantum import QuOperator
+
+
+Gate = gates.Gate
+Tensor = Any
+
+sgates = (
+    ["i", "x", "y", "z", "h", "t", "s", "td", "sd", "wroot"]
+    + ["cnot", "cz", "swap", "cy", "iswap", "ox", "oy", "oz"]
+    + ["toffoli", "fredkin"]
+)
+vgates = [
+    "r",
+    "cr",
+    "rx",
+    "ry",
+    "rz",
+    "rxx",
+    "ryy",
+    "rzz",
+    "crx",
+    "cry",
+    "crz",
+    "orx",
+    "ory",
+    "orz",
+    "any",
+    "exp",
+    "exp1",
+]
+mpogates = ["multicontrol", "mpo"]
+gate_aliases = [
+    ["cnot", "cx"],
+    ["fredkin", "cswap"],
+    ["toffoli", "ccnot"],
+    ["any", "unitary"],
+]
+
+
+class AbstractCircuit:
+    _nqubits: int
+    _qir: List[Dict[str, Any]]
+    split: Optional[Dict[str, Any]]
+    inputs: Tensor
+    circuit_param: Dict[str, Any]
+
+    sgates = sgates
+    vgates = vgates
+    mpogates = mpogates
+    gate_aliases = gate_aliases
+
+    def apply_general_gate(
+        self,
+        gate: Union[Gate, QuOperator],
+        *index: int,
+        name: Optional[str] = None,
+        split: Optional[Dict[str, Any]] = None,
+        mpo: bool = False,
+        ir_dict: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        raise NotImplementedError
+
+    apply = apply_general_gate
```

Comment:
this equation has a problem for inheritance. For child class of abstract class, its ``apply`` method will be redirected to baseclass which is still not implemented

Reply:
also we should add some comments and docstring on what should be handled in this function, apart from gate application we should also take care of qir and so on